core/src/main/java/org/apache/accumulo/core/rpc/clients/TServerClient.java

```diff
@@ -79,22 +85,38 @@ default Pair<String,C> getThriftServerConnection(Logger LOG, ThriftClientTypes<C
     final ZooCache zc = context.getZooCache();
 
     final List<String> serverPaths = new ArrayList<>();
-    zc.getChildren(tserverZooPath).forEach(tserverAddress -> {
-      serverPaths.add(tserverZooPath + "/" + tserverAddress);
-    });
-    if (type == ThriftClientTypes.CLIENT) {
-      zc.getChildren(sserverZooPath).forEach(sserverAddress -> {
-        serverPaths.add(sserverZooPath + "/" + sserverAddress);
-      });
+    if (type == ThriftClientTypes.CLIENT && debugHost != null) {
+      // add all three paths to the set even though they may not be correct.
+      // The entire set will be checked in the code below to validate
+      // that the path is correct and the lock is held and will return the
+      // correct one.
+      serverPaths.add(tserverZooPath + "/" + debugHost);
+      serverPaths.add(sserverZooPath + "/" + debugHost);
       zc.getChildren(compactorZooPath).forEach(compactorGroup -> {
-        zc.getChildren(compactorZooPath + "/" + compactorGroup).forEach(compactorAddress -> {
-          serverPaths.add(compactorZooPath + "/" + compactorGroup + "/" + compactorAddress);
-        });
+        serverPaths.add(compactorZooPath + "/" + compactorGroup + "/" + debugHost);
+      });
+    } else {
+      zc.getChildren(tserverZooPath).forEach(tserverAddress -> {
+        serverPaths.add(tserverZooPath + "/" + tserverAddress);
       });
+      if (type == ThriftClientTypes.CLIENT) {
+        zc.getChildren(sserverZooPath).forEach(sserverAddress -> {
+          serverPaths.add(sserverZooPath + "/" + sserverAddress);
+        });
+        zc.getChildren(compactorZooPath).forEach(compactorGroup -> {
+          zc.getChildren(compactorZooPath + "/" + compactorGroup).forEach(compactorAddress -> {
+            serverPaths.add(compactorZooPath + "/" + compactorGroup + "/" + compactorAddress);
+          });
+        });
+      }
     }
 
     if (serverPaths.isEmpty()) {
-      if (warned.compareAndSet(false, true)) {
+      if (type == ThriftClientTypes.CLIENT && debugHost != null) {
+        LOG.error(
```

Comment:
Not sure if this line is reachable.  To get to this line serverPaths has to be empty, but when debugHost is not null it seems like something is always added to serverPaths earlier in the method.

Reply:
Good catch. Updated code.